pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java

```diff
@@ -76,43 +77,131 @@ public BlobInputStream(LargeObject lo, int bsize) {
    */
   public BlobInputStream(LargeObject lo, int bsize, long limit) {
     this.lo = lo;
-    buffer = null;
-    bpos = 0;
-    apos = 0;
-    this.bsize = bsize;
-    this.limit = limit;
+    this.bufferSize = bsize;
+    // Treat -1 as no limit for backward compatibility
+    this.limit = limit == -1 ? Long.MAX_VALUE : limit;
   }
 
   /**
    * The minimum required to implement input stream.
    */
   public int read() throws java.io.IOException {
-    LargeObject lo = getLo();
-    try {
-      if (limit > 0 && apos >= limit) {
+    try (ResourceLock ignore = lock.obtain()) {
+      LargeObject lo = getLo();
+      if (absolutePosition >= limit) {
         return -1;
       }
-      if (buffer == null || bpos >= buffer.length) {
-        buffer = lo.read(bsize);
-        bpos = 0;
+      // read more in if necessary
+      if (buffer == null) {
+        // Don't hold the buffer while waiting for DB to respond
+        // Note: lo.read(...) does not support "fetching the response into the user-provided buffer"
+        // See https://github.com/pgjdbc/pgjdbc/issues/3043
+        buffer = null;
+        buffer = lo.read(bufferSize);
+        bufferPosition = 0;
       }
 
       // Handle EOF
-      if (buffer == null || bpos >= buffer.length) {
+      if (buffer == null || bufferPosition >= buffer.length) {
         return -1;
       }
 
-      int ret = (buffer[bpos] & 0x7F);
-      if ((buffer[bpos] & 0x80) == 0x80) {
-        ret |= 0x80;
-      }
+      int ret = buffer[bufferPosition] & 0xFF;
 
-      bpos++;
-      apos++;
+      bufferPosition++;
+      absolutePosition++;
+      if (bufferPosition >= buffer.length) {
+        // TODO: support buffer reuse in mark/reset
+        buffer = null;
+        bufferPosition = 0;
+      }
 
       return ret;
-    } catch (SQLException se) {
-      throw new IOException(se.toString());
+    } catch (SQLException e) {
+      long loId = lo == null ? -1 : lo.getLongOID();
+      throw new IOException(
+          GT.tr("Can not read data from large object {0}, position: {1}, buffer size: {2}",
+              loId, absolutePosition, bufferSize),
+          e);
+    }
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    try (ResourceLock ignore = lock.obtain()) {
```

Comment:
b is not a particularly expressive name here. dest would be better.

Reply:
Well, `b` is the canonical name declared in the JDK: https://github.com/openjdk/jdk/blob/abf2e4909733550c228bd66ccebe33a492269e7f/src/java.base/share/classes/java/io/InputStream.java#L281

Reply:
I'm not going to argue strongly about this, but we really don't have to follow their bad practices.